project/02-Modularity/Handout.md

```diff
@@ -0,0 +1,160 @@
+connection.setReadTimeout(3000);
+connection.setConnectTimeout(3000);
+connection.setRequestMethod("GET");
+
+// Already true by default but setting just in case; needs to be true since this request
+// is carrying an input (response) body.
+connection.setDoInput(true);
+
+int responseCode = connection.getResponseCode();
+// Do something with responseCode
+
+stream = connection.getInputStream();
+
+// Do something with the stream
+
+stream.close();
+connection.disconnect();
+```
+>This example is simplified to be the most concise so it is not perfect, it gives a sense of how to perform requests. Feel free to make it great.
+### JSON Parsing
+
+After receiving the result of the HTTP request, you will need to parse it to extract the values and build your data objects.
+
+You should have received the response as a `String`. To make it exploitable, you will want to have a `JSONObject`. Look at the documentation of the class `org.json.JSONTokener` to see how you can do that.
+
+`JSONObject` makes it very simple to access any property in the JSON data you received.
+
+For example, in the following (example) JSON object:
+
+```json
+{
+  "weather": {
+    "temperature": {
+      "current": 10.0
+    }
+  }
+}
+```
+
+To access the temperature in this **example** JSON, you will write something like:
+
+```java
+public double getTemperature(JSONObject json) throws JSONException {
+  return json.getJSONObject("weather").getJSONObject("temperature").getDouble("current");
+}
+```
+
+You will find the exact content of the JSON in the [OneCall API documentation](https://openweathermap.org/api/one-call-api). Please note that in the array `daily`, the first entry (e.g. `daily[0]`) corresponds to the current day.
+
+### Location
+
+To get the location, Android provides a system service called `LocationManager`. To retrieve it, a `Context` is required:
+`(LocationManager) context.getSystemService(Context.LOCATION_SERVICE)`.
+
+This service provides a method called `getLastKnownLocation(...)` (see [the documentation](https://developer.android.com/reference/android/location/LocationManager.html#getLastKnownLocation%28java.lang.String%29)) that immediately returns the last recorded location. To work, it requires a _location provider_ which you can obtain from the method `getBestProvider` of the same class. This method in turn requires a `Criteria` to compare the providers.
+
+> :warning: **Warning:** `getLastKnownLocation` will not return anything if no app tried to get a position before you do. As we are not trying to teach you everything about Android here, this will suffice. To generate a location, you will therefore need to open Google Maps once on the virtual device before actually testing the app. If you want, you can try to use more reliable positionning methods. Feel free!
+
+### GeoCoding
+
+Android provides a small service called [Geocoder](https://developer.android.com/reference/android/location/Geocoder). 
```

Comment:
if Android uses the "Geocoder" spelling with lowercase c, I suggest using that throughout the handout.